### PR TITLE
Use common binary build matrix for binary workflows, Nova and validation workflows

### PR DESCRIPTION
**Pull request by atalman**
Use common binary build matrix for binary workflows, Nova and validation workflows.
 
Test Plan: Run ./regenerate.sh and observe no changes in generated-* workflow files

Issues with current workflows are found and corrected by this refactoring:
1 .github/workflows/generated-linux-binary-libtorch-pre-cxx11-master.yml
2 .github/workflows/generated-macos-arm64-binary-wheel-nightly.yml

This PR just consolidates the matrix, does not change the binary workflow generation procedure.

### CODE REVIEW

File: .github/scripts/generate_ci_workflows.py [hunk 1]
```diff
@@ -83,11 +92,39 @@ class OperatingSystem:
...
+    import importlib.util
+    spec = importlib.util.spec_from_file_location(modname, fname)
+    if spec is None:
+        raise ImportError(f"Could not load spec for module '{modname}' at: {fname}. \
+        Please run the regenerate.sh from .github folder before running generate_ci_workflows")
+    module = importlib.util.module_from_spec(spec)
+    try:
+        assert spec.loader is not None
+        assert module is not None
+        sys.modules[modname] = module
```
Comment (malfet) [inline .github/scripts/generate_ci_workflows.py:105]: Why does it needs to be registered globally?

**Review by malfet — COMMENTED** [review]
This feels needlessly fragile. Why build matrix is not a json file that can be fetched the repo? Alternatively, one can simply execute downloaded script and parse its output? (though from security point of view this feels somewhat unsafe)

**Comment by atalman** [comment]
> Alternatively, one can simply execute downloaded script and parse its output

Let me do this. Does executing script in this way would be good idea ?
```
out = generate_binary_build_matrix.main(["--operating-system", OperatingSystem.LINUX, "--package-type", "manywheel"])
```

File: .github/regenerate.sh [hunk 2]
```diff
@@ -1,6 +1,10 @@
 #!/bin/bash -e
 
+
 # Allows this script to be invoked from any directory:
 cd "$(dirname "$0")"
 
+# Require common generate binary build matrix from test-infra repo
+curl -o scripts/.tools/generate_binary_build_matrix.py --create-dirs https://raw.githubusercontent.com/pytorch/test-infra/main/tools/scripts/generate_binary_build_matrix.py
+
```
[the remark this answers is not in the document]
Comment (atalman) [reply .github/regenerate.sh:9]: I think we do need new version, don't we ? if the underlining script changes, I think we need to update the matrix

File: .github/scripts/generate_ci_workflows.py (hunk 1 above)
Comment (atalman) [reply .github/scripts/generate_ci_workflows.py:121]: I am getting following error without registration:
```
Traceback (most recent call last):
  File "generate_ci_workflows.py", line 130, in <module>
    generate_binary_build_matrix = import_module(
  File "generate_ci_workflows.py", line 124, in import_module
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File ".tools/generate_binary_build_matrix.py", line 21, in <module>
    mod = sys.modules[__name__]
KeyError: 'generate_binary_build_matrix'
```
Python version : Python 3.8.6
This conforms with the doc: https://docs.python.org/3.8/library/importlib.html#importing-a-source-file-directly

File: .github/scripts/generate_ci_workflows.py [hunk 3]
```diff
@@ -107,7 +148,11 @@ class OperatingSystem:
         package_type="libtorch",
         abi_version=generate_binary_build_matrix.CXX11_ABI,
         build_configs=generate_binary_build_matrix.generate_libtorch_matrix(
-            OperatingSystem.LINUX, generate_binary_build_matrix.CXX11_ABI
+            OperatingSystem.LINUX,
+            CHANNEL,
+            with_cuda=generate_binary_build_matrix.ENABLE,
+            with_py311=generate_binary_build_matrix.DISABLE,
```
[the remark this answers is not in the document]
Comment (atalman) [reply .github/scripts/generate_ci_workflows.py]: You are correct, these are not needed. Done.

File: .github/regenerate.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (malfet) [reply .github/regenerate.sh:9]: I though about something along those lines (calling shell scripts from python is indeed ugly):
```python
from urllib.requests import urlopen

def download_url(url: str, fname: Optional[str] = None):
   if fname is None:
       fname = url.rstrip("/", 1)[1]
  with urlopen(url) as conn:
    with open(fname, "wb") as f:
        f.write(conn.read())
```

[the remark this answers is not in the document]
Comment (atalman) [reply .github/regenerate.sh:9]: Ok let me do this! Thanks for code snippet!

File: .github/scripts/generate_ci_workflows.py [hunk 4]
```diff
@@ -77,50 +96,92 @@ def generate_workflow_file(self, workflow_template: jinja2.Template) -> None:
...
+        assert module is not None
+        sys.modules[fname.stem] = module
+        spec.loader.exec_module(module)
+    except FileNotFoundError as e:
+        raise ImportError(f"{e.strerror}: {fname}") from e
+    return module
+
+generate_binary_build_matrix = import_module(
+    Path(__file__).parent / ".tools" / "generate_binary_build_matrix.py"
+)
```
Comment (malfet) [inline .github/scripts/generate_ci_workflows.py:128]: Why this is needed?

File: .github/scripts/generate_ci_workflows.py [hunk 5]
```diff
@@ -77,50 +96,92 @@ def generate_workflow_file(self, workflow_template: jinja2.Template) -> None:
...
+    bin_bld_matrix[osys] = {}
+    for package in PackageType:
+        command = ["--channel", CHANNEL, "--operating-system", osys.value, "--package-type", package.value]
+        if osys == OperatingSystem.MACOS_ARM64 and package == PackageType.LIBTORCH:
+            continue
+        elif osys == OperatingSystem.LINUX and package == PackageType.WHEEL:
+            command += ["--with-py311", ENABLE, "--with-pypi-cudnn", ENABLE]
+        f = io.StringIO()
+        with redirect_stdout(f):
+            generate_binary_build_matrix.main(command)
```
Comment (malfet) [inline .github/scripts/generate_ci_workflows.py:141]:
```suggestion
        f = subprocess.check_output([sys.executable, "generate_binary_build_matrix.py"] + command)
```

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not import the code you've downloaded as submodule, but rather call it as external process and check its output. And why one needs to pass all those extra options? I.e. wouldn't it be less prone to invoke script without any argument, that would presumably return the whole matrix and then filter out the configs we care about.

I.e. if somebody modifies the script to remove `--with-py311` option your program as it is written right now will fail, but in the scenario I propose it will not generate any binary builds, which is probably more desirable.


But just out of curiousity: do other nova projects have to jump thru the same hoops or do they call some sort of reusable actions to return matrix to them. If this is the case, why the same logic can not be applied to core

**Comment by atalman** [comment]
> Please do not import the code you've downloaded as submodule, but rather call it as external process and check its output. And why one needs to pass all those extra options? I.e. wouldn't it be less prone to invoke script without any argument, that would presumably return the whole matrix and then filter out the configs we care about.
> 
> I.e. if somebody modifies the script to remove `--with-py311` option your program as it is written right now will fail, but in the scenario I propose it will not generate any binary builds, which is probably more desirable.
> 
> But just out of curiousity: do other nova projects have to jump thru the same hoops or do they call some sort of reusable actions to return matrix to them. If this is the case, why the same logic can not be applied to core

@malfet as discussed rewriting and automating nightly binaries generation is outside of the scope of this PR. This is meant to be just to unify release matrix across projects. But perhaps you are right, we should rewrite the binary builds workflows altogether. This however inreases a scope of this quite a bit.

**Pull request closed without merging**